src/platform/plugins/shared/discover/public/application/main/components/layout/cascaded_documents/blocks/use_row_header_components.tsx

```diff
@@ -0,0 +1,523 @@
+          return null;
+        }
+
+        popoverRef.current = e.currentTarget as HTMLButtonElement;
+        return this;
+      });
+    },
+    [setPopoverRowData]
+  );
+
+  /**
+   * Renders the popover for the row action (3 dots) button.
+   * Adopting this pattern avoids rendering multiple popovers
+   * in the table which can be very expensive.
+   *
+   * Note: The popover content is re-created each
+   * time it is opened to ensure the context actions have access to the
+   * latest props and state.
+   */
+  const renderRowActionPopover = useCallback(
+    (container?: HTMLElement) => {
+      return popoverRowData ? (
+        <EuiWrappingPopover
+          button={popoverRef.current!}
+          isOpen={popoverRowData !== null}
+          closePopover={() => setPopoverRowData(null)}
+          panelPaddingSize="none"
+          anchorPosition="upLeft"
+          container={container}
+        >
+          <ContextMenu
+            close={closePopover}
+            editorQuery={editorQuery}
+            esqlVariables={esqlVariables}
+            globalState={globalState}
+            row={popoverRowData}
+            services={services}
+            dataView={dataView}
+            statsFieldSummary={statsFieldSummary}
+            openInNewTab={openInNewTab}
+            updateESQLQuery={updateESQLQuery}
+          />
+        </EuiWrappingPopover>
+      ) : null;
+    },
+    [
+      popoverRowData,
+      closePopover,
+      editorQuery,
+      esqlVariables,
+      globalState,
+      services,
+      dataView,
+      statsFieldSummary,
+      openInNewTab,
+      updateESQLQuery,
+    ]
+  );
+
+  return {
+    renderRowActionPopover,
+    togglePopover,
+  };
+};
+
+const textSlotStyles = {
+  textWrapper: css({
+    minWidth: 0,
+    textWrap: 'nowrap',
+  }),
+  textInner: css({
+    overflow: 'hidden',
+    textOverflow: 'ellipsis',
+  }),
+};
+
+export function useEsqlDataCascadeRowHeaderComponents(
+  editorQueryMeta: ESQLStatsQueryMeta,
+  selectedColumns: string[],
+  togglePopover: ReturnType<typeof useEsqlDataCascadeRowActionHelpers>['togglePopover']
+) {
+  const namedColumnsFromQuery = useMemo(() => {
+    return editorQueryMeta.appliedFunctions.map(({ identifier }) => identifier);
+  }, [editorQueryMeta.appliedFunctions]);
+
+  /**
+   * Renders the title part of the row header.
+   */
+  const rowHeaderTitle = useCallback<
+    NonNullable<DataCascadeRowProps<ESQLDataGroupNode, DataTableRecord>['rowHeaderTitleSlot']>
+  >(
+    ({ rowData, nodePath }) => {
+      const rowGroup = nodePath[nodePath.length - 1];
+      const type = editorQueryMeta.groupByFields.find((field) => field.field === rowGroup)?.type;
+
+      if (type && /categorize/i.test(type)) {
+        return (
+          <div data-test-subj={`${rowData.id}-dscCascadeRowTitlePatternCellRenderer`}>
+            {getPatternCellRenderer(
+              // @ts-expect-error - necessary to match the data shape expectation
```

Comment:
I'm not a fan of using `@ts-expect-error` here. Why don't we have access to the actual `DataTableRecord`?